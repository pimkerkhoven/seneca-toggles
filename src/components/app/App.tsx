import React, {useState} from 'react';
import { Question } from "../../types/Question";
import TogglesQuestion from "../toggles-question/TogglesQuestion";

function App() {
  const questions: Question[] = [{
    title: "An animal cell contains:",
    parts: [
      {
        options: ["Cell wall", "Ribosomes", "Glucose", "Printers"],
        answer: "Cell wall"
      },
      {
        options: ["Cytoplasm", "Chloroplast"],
        answer: "Cytoplasm"
      },
      {
        options: ["Partial membrane", "Impermeable membrane"],
        answer: "Impermeable membrane"
      }
    ]
  }, {
    title: "Things I can see right now:",
    parts: [
      {
        options: ["Cat", "Dog"],
        answer: "Cat"
      },
      {
        options: ["Ecology book", "Geology book"],
        answer: "Ecology book"
      },
      {
        options: ["Apple", "Banana", "Watermelon"],
        answer: "Apple"
      },
      {
        options: ["Glass", "Laptop charger"],
        answer: "Glass"
      }
    ]
  }]

  const [activeQuestionIndex, setActiveQuestionIndex] = useState<number>(0)

  function nextQuestion(event: React.MouseEvent) {
    event.preventDefault()

    setActiveQuestionIndex(prevState => (prevState + 1) % questions.length)
  }

  return <div style={{maxWidth: 900, margin: "0 auto"}}>
    <button onClick={nextQuestion}>Next Question</button>
    <TogglesQuestion question={questions[activeQuestionIndex]} />
  </div>
}

export default App;
