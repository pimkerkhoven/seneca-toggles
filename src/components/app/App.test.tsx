import React from 'react';
import {fireEvent, render, screen} from '@testing-library/react';
import App from './App';


/**
 * - test snapshot of initial render
 * - test clicking answer toggles option
 * - test putting correct answer updates status text
 * - test putting correct answer locks toggles
 */

// TODO: move tests to correct file!

test('renders the component', () => {
  const { asFragment } = render(<App />);
  expect(asFragment()).toMatchSnapshot();
});

test('toggling an option marks it as selected', () => {
  const {asFragment} = render(<App />);

  const toggle = screen.getByText("Ribosomes")
  fireEvent.click(toggle)

  expect(asFragment()).toMatchSnapshot()
  expect(toggle.className).toBe( "ToggleOption selected")
  expect(screen.getByText("Cell wall").className).toBe("ToggleOption")
});

test('marking last incorrect answer correct updates status text', () => {
  const { asFragment } = render(<App />);

  const statusText = screen.getByText("The answer is incorrect")

  const toggle = screen.getByText("Impermeable membrane")
  fireEvent.click(toggle)

  expect(asFragment()).toMatchSnapshot()
  expect(statusText.innerHTML).toBe("The answer is correct!")
});

test('marking last incorrect answer correct locks toggles', () => {
  const { asFragment } = render(<App />);

  const toggle = screen.getByText("Impermeable membrane")
  fireEvent.click(toggle)

  const lockedToggle = screen.getByText("Ribosomes")
  fireEvent.click(lockedToggle)

  expect(asFragment()).toMatchSnapshot()
  expect(lockedToggle.className).toBe("ToggleOption")
});


