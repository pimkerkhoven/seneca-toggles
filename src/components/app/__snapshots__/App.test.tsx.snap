// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`marking last incorrect answer correct locks toggles 1`] = `
<DocumentFragment>
  <div
    class="TogglesQuestion"
  >
    <h1>
      An animal cell contains:
    </h1>
    <div
      class="Toggle"
    >
      <div
        class="ToggleOption selected"
      >
        Cell wall
      </div>
      <div
        class="ToggleOption"
      >
        Ribosomes
      </div>
    </div>
    <div
      class="Toggle"
    >
      <div
        class="ToggleOption selected"
      >
        Cytoplasm
      </div>
      <div
        class="ToggleOption"
      >
        Chloroplast
      </div>
    </div>
    <div
      class="Toggle"
    >
      <div
        class="ToggleOption"
      >
        Partial membrane
      </div>
      <div
        class="ToggleOption selected"
      >
        Impermeable membrane
      </div>
    </div>
    <h2>
      The answer is correct!
    </h2>
  </div>
</DocumentFragment>
`;

exports[`marking last incorrect answer correct updates status text 1`] = `
<DocumentFragment>
  <div
    class="TogglesQuestion"
  >
    <h1>
      An animal cell contains:
    </h1>
    <div
      class="Toggle"
    >
      <div
        class="ToggleOption selected"
      >
        Cell wall
      </div>
      <div
        class="ToggleOption"
      >
        Ribosomes
      </div>
    </div>
    <div
      class="Toggle"
    >
      <div
        class="ToggleOption selected"
      >
        Cytoplasm
      </div>
      <div
        class="ToggleOption"
      >
        Chloroplast
      </div>
    </div>
    <div
      class="Toggle"
    >
      <div
        class="ToggleOption"
      >
        Partial membrane
      </div>
      <div
        class="ToggleOption selected"
      >
        Impermeable membrane
      </div>
    </div>
    <h2>
      The answer is correct!
    </h2>
  </div>
</DocumentFragment>
`;

exports[`renders the component 1`] = `
<DocumentFragment>
  <div
    class="TogglesQuestion"
  >
    <h1>
      An animal cell contains:
    </h1>
    <div
      class="Toggle"
    >
      <div
        class="ToggleOption selected"
      >
        Cell wall
      </div>
      <div
        class="ToggleOption"
      >
        Ribosomes
      </div>
    </div>
    <div
      class="Toggle"
    >
      <div
        class="ToggleOption selected"
      >
        Cytoplasm
      </div>
      <div
        class="ToggleOption"
      >
        Chloroplast
      </div>
    </div>
    <div
      class="Toggle"
    >
      <div
        class="ToggleOption selected"
      >
        Partial membrane
      </div>
      <div
        class="ToggleOption"
      >
        Impermeable membrane
      </div>
    </div>
    <h2>
      The answer is incorrect
    </h2>
  </div>
</DocumentFragment>
`;

exports[`toggling an option marks it as selected 1`] = `
<DocumentFragment>
  <div
    class="TogglesQuestion"
  >
    <h1>
      An animal cell contains:
    </h1>
    <div
      class="Toggle"
    >
      <div
        class="ToggleOption"
      >
        Cell wall
      </div>
      <div
        class="ToggleOption selected"
      >
        Ribosomes
      </div>
    </div>
    <div
      class="Toggle"
    >
      <div
        class="ToggleOption selected"
      >
        Cytoplasm
      </div>
      <div
        class="ToggleOption"
      >
        Chloroplast
      </div>
    </div>
    <div
      class="Toggle"
    >
      <div
        class="ToggleOption selected"
      >
        Partial membrane
      </div>
      <div
        class="ToggleOption"
      >
        Impermeable membrane
      </div>
    </div>
    <h2>
      The answer is incorrect
    </h2>
  </div>
</DocumentFragment>
`;
